controllers/metal3.io/dataimage_controller.go

```diff
@@ -0,0 +1,265 @@
+//
+// For more details, check Reconcile and its Result here:
+// - https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.14.4/pkg/reconcile
+func (r *DataImageReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	reqLogger := r.Log.WithValues("dataimage", req.NamespacedName)
+	reqLogger.Info("start dataImage reconciliation V1")
+
+	di := &metal3api.DataImage{}
+	if err := r.Get(ctx, req.NamespacedName, di); err != nil {
+		// The DataImage resource may have been deleted
+		if k8serrors.IsNotFound(err) {
+			reqLogger.Info("dataImage not found")
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{Requeue: true, RequeueAfter: dataImageRetryDelay}, fmt.Errorf("could not load dataImage, %w", err)
+	}
+
+	bmh := &metal3api.BareMetalHost{}
+	if err := r.Get(ctx, req.NamespacedName, bmh); err != nil {
+		// There might not be any BareMetalHost for the DataImage
+		if k8serrors.IsNotFound(err) {
+			reqLogger.Info("bareMetalHost not found for the dataImage")
+			return ctrl.Result{}, nil
+		}
+
+		// Error reading the object - requeue the request.
+		return ctrl.Result{Requeue: true, RequeueAfter: dataImageRetryDelay}, fmt.Errorf("could not load baremetalhost, %w", err)
+	}
+
+	info := &rdiInfo{ctx: ctx, log: reqLogger, request: req, di: di, bmh: bmh}
+
+	if hasDetachedAnnotation(bmh) {
+		reqLogger.Info("the host is detached, not running reconciler")
+		return ctrl.Result{Requeue: true, RequeueAfter: unmanagedRetryDelay}, nil
+	}
+
+	// If the reconciliation is paused, requeue
+	annotations := bmh.GetAnnotations()
+	if annotations != nil {
+		if _, ok := annotations[metal3api.PausedAnnotation]; ok {
+			reqLogger.Info("host is paused, no work to do")
+			return ctrl.Result{Requeue: false}, nil
+		}
+	}
+
+	// Create a provisioner that can access Ironic API
+	prov, err := r.ProvisionerFactory.NewProvisioner(ctx, provisioner.BuildHostDataNoBMC(*bmh), info.publishEvent)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to create provisioner, %w", err)
+	}
+
+	ready, err := prov.TryInit()
+	if err != nil || !ready {
+		var msg string
+		if err == nil {
+			msg = "Not ready"
+		} else {
+			msg = err.Error()
+		}
+		reqLogger.Info("provisioner is not ready", "Error", msg, "RequeueAfter", provisionerRetryDelay)
+		return ctrl.Result{Requeue: true, RequeueAfter: provisionerRetryDelay}, nil
+	}
+
+	// Add finalizer for newly created DataImage
+	if di.DeletionTimestamp.IsZero() && !utils.StringInList(di.Finalizers, metal3api.DataImageFinalizer) {
+		reqLogger.Info("adding finalizer")
+		di.Finalizers = append(di.Finalizers, metal3api.DataImageFinalizer)
+
+		// Update dataImage after adding finalizer, requeue in case of failure
+		err := r.Update(ctx, di)
+		if err != nil {
+			return ctrl.Result{RequeueAfter: dataImageUpdateDelay}, fmt.Errorf("failed to update resource after add finalizer, %w", err)
+		}
+		return ctrl.Result{Requeue: true}, nil
+	}
+
```

Comment:
nit: here you can use the newly introduced HasDataImage to provide a better error message when Ironic is too old